### DISCUSSION TITLE
Linear constant velocity of the icub hand

### DISCUSSION THREAD

**Discussion by tasbolat1** [robotology/community#291] [Q&A] [answered]
Hello, everyone!
I want the hand of icub to move with constant velocity along `y` axis. To do so, I used Cartesian Control and with the following settings:
```

xdot[0]=0.0; xdot[1]=-0.003; xdot[2]=0.0;
odot = 0.0;
icart->setTaskVelocities(xdot, odot);
```

where `icart` is `ICartesianControl`. I haven't set any other configurations (no trajectory time). And, asked the icub hand to move to desired position like this:

icart->goToPose(xd,od);

where hand moves form  (-0.3, 0, 0.1) to (-0.3. -0.3, 0.1). However, the icub simulation result doesn't correspond to what I asked. Here are figures:
Motion along `y` axis:
![xd1](https://user-images.githubusercontent.com/4617666/40093415-bb9d3346-58f4-11e8-969c-0d9de572b99e.png)

Motion along `x` axis
![xd0](https://user-images.githubusercontent.com/4617666/40093417-bdfd10ca-58f4-11e8-9d79-6c3655189129.png)

Motion along `z` axis
![xd2](https://user-images.githubusercontent.com/4617666/40093420-c4e377a8-58f4-11e8-979d-b82d8cbd191b.png)

about figures:
 - we expect that there's no change (at least very small) in `x` and `z` axes. However, they change more than 4 cm and 1 cm respectively,  which, I think it's too big.
- in the first figure (along `y` axis), the velocity hasn't been set as I set using `setTaskVelocities`.

So, my questions are:
- Where I'm doing wrong?
- Is it, at the end of the day, feasible with Icub hand to have a linear motion with constant speed?
- If so, please suggest ways to do so.

Thanks, Tasbolat

**Comment by xEnVrE** [comment, marked as answer]
Hi @tasbolat1,

it is indeed possible to have the iCub hand to follow a constant linear velocity trajectory. Let me clarify some points before getting to a possible solution to the problem.

As far as I understand you are trying to accomplish this task by first issuing a `setTaskVelocities` command and then a `goToPose` command. In this case the first command would be **overridden** by the second and the Cartesian Controller will essentially perform a point to point control moving the palm of the hand from the initial position ` (-0.3, 0, 0.1)` to the final position ` (-0.3, -0.3, 0.1)` using a Minimum-Jerk trajectory. This kind of trajectory has several desirable properties but, as you can find out from the figures that you have attached, it does not correspond to a straight line from the initial to the final position as you expected.

In order to achieve your goal, you should issue the `setTaskVelocities` command alone. When such a command is issued, the Controller **integrates** internally the commanded velocity and produces a reference trajectory whose position and orientation are tracked as normally happens when you issue a `goToPose` command - you can imagine this as issuing multiple times the `goToPose` command with intermediate poses along the desired trajectory that is specified in terms of its  **instantaneous** velocity using `setTaskVelocities`.

A very important parameter to be set in order to shape the resulting trajectory is the trajectory time. The trajectory time is not just the time required to move the end-effector from the initial pose to a given final pose. In fact, it determines the reactivity of the Controller, i.e. **lower** values of the trajectory time determines **faster** movements of the robot and vice-versa. This also applies to the `setTaskVelocities` command since, as I explained, it essentially produces several intermediate poses that will be tracked according to the specified trajectory time.

I would suggest you read #208 that provides useful insights on how to tune the trajectory time when performing control with `setTaskVelocities` and also some [example code](https://gist.github.com/pattacini/41355442915d11b7b77eb46770f3df52).

Finally, please note that, although you are not setting a trajectory time, a default value of `2.0s` is set within the Controller.

**Comment by tasbolat1** [comment]
@xEnVrE , thank you very much directing me to the place where I could find useful informations!

Regarding the use of `setTaskVelocities`, I have some questions.

The problem statement (for now, everything is done in simulation):
I still need to have a constant linear velocity between `(-0.3, 0, 0.1)` and `(-0.3, -0.25, 0.1)`. The displacement has been shrunk a little bit to ensure that those points are reachable. Using `setTaskVelocities`, I can get the following result for different trajectory time:

Parameters used in simulation:
`Update_period = 0.02 [s]`
`Velocity = -0.03 [m/s]` along y-axis


`Traj_time == 0.1 s:`
![y_axis_0_1](https://user-images.githubusercontent.com/4617666/40365180-9aa8c490-5e06-11e8-86e1-3d971e9543ba.png)


`Traj_time == 0.4 s:`
![y_axis_0_4](https://user-images.githubusercontent.com/4617666/40365189-9ddd5054-5e06-11e8-958e-be5eac5e4be7.png)


`Traj_time == 0.7 s:`
![y_axis_0_7](https://user-images.githubusercontent.com/4617666/40365192-a04f824e-5e06-11e8-82d7-ee96883d73e2.png)


`Traj_time == 1 s:`
![y_axis_1](https://user-images.githubusercontent.com/4617666/40365206-a32edafa-5e06-11e8-8555-d97adcec9a2e.png)


About figures:
`red = x, green = y, blue = z`
The velocity profile data has been recorded by `getTaskVelocities` during each iteration of loop.

The questions are:
1. Can we have more smooth velocity profile than given in the figures? If so, which parameter should be tuned?
2. When the hand reaches the final point, 0 velocity is set using `setTaskVelocities`. However, as seen in figures, it sharply stops the motion. This case may damage real robot (it's my guess).
3. In real robot, how much external friction will have an effect?

Thanks,
Tasbolat

**Comment by pattacini** [comment]
Hi @tasbolat1 

As @xEnVrE explained with plenty of details, you cannot assign a Cartesian velocity straight to the end-effector. Rather, you can instruct the controller to strive to track a target point moving at the specified velocity. Thus, the controller will respond with its own dynamics (look up this [paper][1]), which has not been purposely designed to generate constant velocity in the operational space. This is essentially why you're seeing a lot of variability in the velocity you dumped.

What you get by yielding `getTaskVelocities` represents the result of the following operations:
1. Access the joint velocities command sent to the robot, say `vel_jnt_cmd`.
1. Multiply this command by the Jacobian to get the corresponding Cartesian velocity of the end-effector, say `vel_ee`.

Since there exists a control loop running underneath, `vel_ee` is the predicted velocity, not the real feedback, which in turn can be better estimated by reading back the Cartesian position (use `getPose`) to then extract the relative speed in a robust manner (e.g. you may want to use this [component][2]).

To sum up, don't expect to get smoother velocities other than keeping fiddling with the parameters we already pointed out. To obtain constant velocities, indeed, the architecture would need to be modified significantly. However, even though at a very detailed observation the resulting velocity isn't constant at all, the visible effect at macroscopic level turns out to be quite acceptable (as can been seen in #208). 

Regarding the abrupt stop, you are meant to be responsible for providing an appropriate velocity profile that embeds a gentler stopping phase. So, in practice, you have to foresee the proper logic within your user code. That said, a snap stop is not very problematic with the real robot.

Your last question about the friction is a bit too generic, very task specific and is also out of my knowledge 😉 

[1]: https://ieeexplore.ieee.org/document/5650851
[2]: http://wiki.icub.org/iCub/main/dox/html/classiCub_1_1ctrl_1_1AWLinEstimator.html

**Comment by tasbolat1** [comment]
Thanks @pattacini  and @xEnVrE for the help. I could manage to achieve constant linear velocity to the end-effector using above explained methods in real robot. Of course, the velocity is not absolutely constant, but it's acceptable in my task.
Thanks again,
Tasbolat

**Comment by pattacini** [comment]
Happy to have helped you out 👍
